### ISSUE TITLE
Missing lazy initialization for tangential process during dataset splitting

### ISSUE THREAD

**Issue by nzllim** [PavlidisLab/Gemma#130]
Refer for Log: https://github.com/PavlidisLab/GemmaCuration/issues/13#issuecomment-770518435

Error was raised during dataset splitting, although the splitting process itself was successful. Non-critical issue related to missing lazy initialization of roles.

**Comment by arteymix** [comment]
Yeah, that happens when relationships are queried outside a transaction.

For this specific case, it's best to annotate the split() service method with `@Transactional` because there's a lot of inter-dependent operations. I'll test it out.

**Comment by ppavlidis** [comment]
I don't know a) why this wasn't an issue before and b) why it isn't in a transaction already, but be very careful as simplying adding that can break things. There is a unit test

**Comment by arteymix** [comment]
I'm trying to reproduce the bug, but I think I need a data vectors fixture in the test. 

```
Caused by: ubic.gemma.core.analysis.preprocess.PreprocessingException: java.lang.RuntimeException: java.lang.IllegalArgumentException: No preferred data vectors for ExpressionExperiment Id=6 Name=Shifts in Ribosome Engagement Impact Key Gene Sets in Neurodevelopment and Rett Syndrome Neurons (GSE123753)
	at ubic.gemma.core.analysis.preprocess.PreprocessorServiceImpl.process(PreprocessorServiceImpl.java:125)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at org.springframework.retry.interceptor.RetryOperationsInterceptor$1.doWithRetry(RetryOperationsInterceptor.java:69)
	... 39 more
```

@ppavlidis any suggestion?

**Comment by ppavlidis** [comment]
Yes, it can be a pain to set up tests like this. In the setup() for the new test, you can just skip the preprocessing step, or if you want to make sure you exercise the data parts, you would have to make a small data matrix file to be loaded in (assuming no data is in the GEO record, that should be getting in assuming chopSoft worked correctly)

**Comment by arteymix** [comment]
Good. The test pass now (with @Transactional). I'll try without to make sure this issue can be reproduced.

**Comment by arteymix** [comment]
I had a weird Hibernate issue with `ExpressionExperimentSevice.delete` with a collection of split experiments. It's complaining about non thread-safe access. It works however if parts are deleted one-by-one.

I also added a `postProcess` flag to split to make processing of the resulting split optional, so we can split EE that don't have data vectors.

**Comment by arteymix** [comment]
The other important thing now is that the whole split operation is transactional, so if a failure occurs, there shouldn't be leftovers.

**Comment by arteymix** [comment]
Reopening this because I realized that `PreprocessorService.process` does not work in a transaction because it manages to somehow duplicated the EE in the session (edit: I'm now looking at an ExperimentalFactor duplication). I tracked the issue down to `processDiagnostics` and `processForPca`. 

If this succeeds, many more operations could be marked as transactional, making Gemma far more robust to errors. The contingency plan is to revert the Transactional annotation and just perform the otherParts in a transaction.
